rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -277,15 +311,121 @@ typedef void (* signal_handler_t)(int);
 signal_handler_t rclcpp::SignalHandler::old_signal_handler = 0;
 #endif
 
-/// Mutex to protect g_sigint_guard_cond_handles
-static std::mutex g_sigint_guard_cond_handles_mutex;
-/// Guard conditions for interrupting the rmw implementation when the global interrupt signal fired.
-static std::map<rcl_wait_set_t *, rcl_guard_condition_t> g_sigint_guard_cond_handles;
+#if defined(_WIN32)
+#include <windows.h>
+#elif defined(__APPLE__)
+#include <dispatch/dispatch.h>
+#else  // posix
+#include <semaphore.h>
+#endif
+
+static std::atomic_bool g_wait_for_signal_is_setup = ATOMIC_VAR_INIT(false);
+#if defined(_WIN32)
+HANDLE g_signal_handler_sem;
+#elif defined(__APPLE__)
+dispatch_semaphore_t g_signal_handler_sem;
+#else  // posix
+sem_t g_signal_handler_sem;
+#endif
 
-/// Condition variable for timed sleep (see sleep_for).
-static std::condition_variable g_interrupt_condition_variable;
-/// Mutex for protecting the global condition variable.
-static std::mutex g_interrupt_mutex;
+void
+rclcpp::setup_wait_for_signal()
+{
+#if defined(_WIN32)
+  g_signal_handler_sem = CreateSemaphore(
+    NULL,  // default security attributes
+    0,  // initial semaphore count
+    1,  // maximum semaphore count
+    NULL);  // unnamed semaphore
+  if (NULL == g_signal_handler_sem) {
+    throw std::runtime_error("CreateSemaphore() failed in setup_wait_for_signal()");
+  }
+#elif defined(__APPLE__)
+  g_signal_handler_sem = dispatch_semaphore_create(0);
+#else  // posix
+  if (-1 == sem_init(&g_signal_handler_sem, 0, 0)) {
+    throw std::runtime_error(std::string("sem_init() failed: ") + strerror(errno));
+  }
+#endif
+  g_wait_for_signal_is_setup.store(true);
+}
+
+void
+rclcpp::teardown_wait_for_signal() noexcept
+{
+  if (!g_wait_for_signal_is_setup.exchange(false)) {
+    return;
+  }
+#if defined(_WIN32)
+  CloseHandle(g_signal_handler_sem);
+#elif defined(__APPLE__)
+  dispatch_release(g_signal_handler_sem);
+#else  // posix
+  if (EINVAL == sem_destroy(&g_signal_handler_sem)) {
+    RCLCPP_ERROR(g_logger, "invalid semaphore in teardown_wait_for_signal()");
+  }
+#endif
+}
+
+void
+rclcpp::wait_for_signal() noexcept
```

Comment:
@wjwwood It seems to me this function could fail and the caller wouldn't know about it. If we want to keep it `noexcept`, consider returning a boolean.

Reply:
I'll just throw in this case, since it's only called from the singleton's thread.